src/main/java/com/_up/megastore/services/implementations/OrderService.java

```diff
@@ -81,9 +110,26 @@ private synchronized Integer getNextOrderNumber() {
         return orderRepository.getLastOrderNumber() + 1;
     }
 
+    private void sendOrderEmail(Order order, String subject) {
+        String emailBody = emailBuilder.buildOrderEmail(order, subject);
+        emailService.sendEmail(order.getUser().getEmail(), subject, emailBody);
+    }
+
     private void throwExceptionIfStateIsNotInProgress(Order order) {
         if (order.getState() != State.IN_PROGRESS) {
             throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Order is not in progress.");
         }
     }
+
+    private void throwExceptionIfStateIsNotFinished(Order order) {
+        if (order.getState() != State.FINISHED) {
+            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Order is not finished.");
+        }
+    }
+
+    private void throwExceptionIfStateIsNotInDelivery(Order order) {
+        if (order.getState() != State.IN_DELIVERY) {
+            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Order is not in delivery.");
+        }
+    }
```

Comment:
Siguiendo los cambios de los enums, esto podriamos modificarlo asi

```suggestion
    private void validatePreviousStates(Order order, State newState) {
        if(newState.previousStates.contains(order.getState())) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Order is not in the correct state.");
        }
    }
```

NOTA: verifica que ande, porque no estoy del todo seguro, pero deberia ir por ahi la cosa